network/kademlia_test.go

```diff
@@ -370,80 +390,110 @@ func binStr(a *BzzAddr) string {
+	tk.Register("01000000", "01100000", "01110000")
 
-	tk.On("01000000")
-	tk.Off("10000001")
-	tk.checkSuggestPeer("10000001", 0, true)
+	//Bins should fill from  most empty to least empty and shallower to deeper
+	//first suggestion should be for bin 0
+	tk.checkSuggestPeer("11111000", 0, false)
+	tk.On("11111000")
 
-	tk.On("00100001")
-	tk.Off("01000000")
-	tk.checkSuggestPeer("01000000", 0, false)
+	//Since we now have 1 peer in bin0 and none in bin1, next suggested peer should be for bin1
+	tk.checkSuggestPeer("01110000", 0, false)
+	tk.On("01110000")
 
-	// second time disconnected peer not callable
-	// with reasonably set Interval
-	tk.checkSuggestPeer("<nil>", 0, false)
+	//Both bins 0 and 1 have at least 1 peer, so next suggested peer should be for 0 (shallower)
+	tk.checkSuggestPeer("11110000", 0, false)
+	tk.On("11110000")
 
-	// on and off again, peer callable again
-	tk.On("01000000")
-	tk.Off("01000000")
-	tk.checkSuggestPeer("01000000", 0, false)
+	//Bin0 has 2 peers, bin1 has 1 peer, should recommend peer for bin 1
+	tk.checkSuggestPeer("01100000", 0, false)
+	tk.On("01100000")
 
-	tk.On("01000000", "10000001")
+	//Bin1 should be saturated now (size == expectedSizeBin1)
+	//Next suggestions should all be bin0 peers
+	tk.checkSuggestPeer("11100000", 0, false)
+	tk.On("11100000")
+	tk.checkSuggestPeer("11000000", 0, false)
+	tk.On("11000000")
+
+	//Bin0 should also  be saturated now (size == expectedSizeBin0)
+	//All bins saturated, shouldn't suggest more peers
 	tk.checkSuggestPeer("<nil>", 0, false)
 
-	tk.Register("00010001")
-	tk.checkSuggestPeer("00010001", 0, false)
+	//We check for change of saturationDepth
+	tk.Off("11000000")
+	tk.checkSuggestPeer("11000000", 0, true)
+	tk.On("11000000")
 
-	tk.On("00010001")
-	tk.Off("01000000")
+	//We check again, this time unsaturating bin1
+	tk.Off("01110000")
+	//We don't connect to force  peer not callable
+	tk.checkSuggestPeer("01110000", 1, true)
 	tk.checkSuggestPeer("01000000", 0, false)
-
-	tk.On("01000000")
+	//Now no peers are callable
 	tk.checkSuggestPeer("<nil>", 0, false)
 
-	tk.Register("01000001")
-	tk.checkSuggestPeer("01000001", 0, false)
-
-	tk.On("01000001")
-	tk.checkSuggestPeer("<nil>", 0, false)
+	tk.On("01110000")
+	tk.Off("01110000")
+	//Peer should be callable again
+	tk.checkSuggestPeer("01110000", 1, true)
+	tk.On("01110000")
 
-	tk.Register("10000010", "01000010", "00100010")
+	//Same with the other peer
+	tk.On("01000000")
+	tk.Off("01000000")
+	//Peer is callable again, but all bins  saturated. No peer suggested
 	tk.checkSuggestPeer("<nil>", 0, false)
 
-	tk.Register("00010010")
-	tk.checkSuggestPeer("00010010", 0, false)
-
-	tk.Off("00100001")
-	tk.checkSuggestPeer("00100010", 2, true)
+	//Depth is 2
+	//Since depth is 2, bins >= 2 aren't considered saturated if peers left to connect
+	//We add addresses that fall in bin2 and bin3
+	tk.Register("00111000", "00110000")
+	tk.Register("00011100")
+
+	//Again, should fill from shallow to deep
```

Comment:
Actually I think this is wrong. Within the neighborhood the closest peers should be connected first.

Reply:
You are refering to the comment?

Reply:
I think the code is doing what the code says? So if that is the case then I believe the sequence of connections within neighborhood are wrong in the implementation.